### PR TITLE
CONTRIBUTING: cleanup and reorganize

### PR DESCRIPTION
**Pull request by silkeh**
### Contribution description

The contributing document was not very concise and to the point. Users often expect a contributing document to explain all required steps of contributing. This set of commits does the following:

  - Link to both the IRC and Matrix end of the chat, as both are used.

  - Replace 'help wanted' by 'contributing code':
    The naming of the 'help wanted' section is unclear, and unhelpful.
    Additionally, most of the items listed in it are not actually part of
    the contribution process:
    
    - While *Documentation* is welcome, it is not helpful to document the
      kinds in the contributing document. The contributing document should
      document *how*, not *what*.
    
    - The *Issues* section is out of place: I doubt anyone will look at the
      *contributing guidelines* to find out how to open an issue.
      The content is also pretty self-explanatory.
    
    - What's left is only 'Contributing Code', so promote it!

  - Remove 'feature request' and 'bug reports' sections
    
    These have fairly clear templates in the GitHub interface.
    Duplicating that here servers no purpose

  - Reorganize 'contributing code' and 'pull request':
    
    Having two lists that need to be read thouroughly for pull requests is
    unclear. Clean this up by:
    
    - Splitting the *pull request* steps into *coding conventions*,
      *commit conventions*, and *pull requests*
    
    - Completing the contribution steps with references to those sections.
    
    - Reordering all steps to any contributor will encounter them in order.
    
    - Rename and promote 'Git usage', to a 'Working with Git' section.
      This is then referred to at the start of the document and in
      the contribution steps.

  - Move the 'writing documentation' section:
    
    Because the *working with git* section is quite long, it obscures the
    *writing documentation* section. Move it to before the *working with
    git* section.

  - Add an example commit to the *commit conventions* section.
    Refer to this in the pull request steps

  - Change the code blocks to triple-backtick (` ``` `) style in order
    to 1) provide a consistent indent and 2) enable syntax highlighting.

 - Add `--no-backup` to uncrustify

### Testing procedure

Read the [rendered README](https://github.com/silkeh/RIOT/blob/pr/cleanup-contributing/CONTRIBUTING.md).

### Issues/PRs references

None.

### CODE REVIEW

**Review by waehlisch — CHANGES_REQUESTED** [review]
> Remove 'feature request' and 'bug reports' sections

having such sections in this document has two advantages: (1) it makes clear that feature requests and bug reports are some kind of contribution to RIOT, which we value. (2) the reader knows where to submit the contribution. 

i would like to see these sections back in the document, maybe Bug Reports a bit shortened.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -257,17 +165,40 @@ into the source repository.
...
+Documentation improvements are always welcome and a good starting point for
+new contributors. This kind of contribution is merged quite quickly in general.
+
+RIOT documentation is built with [doxygen][doxygen]. Doxygen is configured to
+parse header (.h) and `doc.txt` files in the RIOT source code to generate
+the modules, cpus, boards and packages documentation.
+General documentation pages are written in Markdown and located in
+`doc/doxygen/src`.
+
+To generate the documentation, simply run `make doc`
```
Comment (waehlisch) [inline CONTRIBUTING.md:193]:
```
$ make doc
```

is consistent with layout in previous sections and makes it easier for the reader to spot the important commands.

**Comment by silkeh** [comment]
> > Remove 'feature request' and 'bug reports' sections
> 
> having such sections in this document has two advantages: (1) it makes clear that feature requests and bug reports are some kind of contribution to RIOT, which we value. (2) the reader knows where to submit the contribution.
> 
> i would like to see these sections back in the document, maybe Bug Reports a bit shortened.

This requires that people actually the contributing guidelines before (wanting to) submitting a bug report or feature request. I don't think that's the case.

I agree that feature requests and bug reports are important contributions! But maybe a better place for this is the README? That file is a lot more prominent, and commonly read. I can open a PR for this.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -257,17 +165,40 @@ into the source repository.
   area of the PR. You can also advertise the PR on devel@riot-os.org mailing
   list and ask for a review there.
 
+* Try to answer reviews as quickly as possible to speed up the review process
+  and avoid stalled PRs.
+
 You can find more information about RIOT development procedure on this
 [wiki page][development-procedures].
```
Comment (waehlisch) [inline CONTRIBUTING.md:172]: this link was and still is broken.

**Review by waehlisch — COMMENTED** [review]
> This requires that people actually the contributing guidelines before (wanting to) submitting a bug report or feature request. I don't think that's the case.
> 
i didn't do any research on what is more likely. but if we have a file that says CONTRIBUTING, i expect that the information is complete.

> I agree that feature requests and bug reports are important contributions! But maybe a better place for this is the README? That file is a lot more prominent, and commonly read. I can open a PR for this.

let's wait for the other reviewers.

File: CONTRIBUTING.md (hunk 2 above)
Comment (silkeh) [reply CONTRIBUTING.md:172]: If you have the right link I'll open a separate PR fixing it!

**Comment by aabadie** [comment]
> Remove 'feature request' and 'bug reports' sections

These 2 sections were here to make sure contributors have checked that the same pull request wasn't already opened before opening one and that no report for the same bug already exists.
I agree that they could be simplified, shortened but not completely removed.

Note that I initially rewrote this document in #10029 by taking inspiration from the [rust contributing document](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md). In the meantime, Rust has completely refactored it: their document now points to their related documentation.

I think we should do the same for RIOT: move details to doc.riot-os.org (in different pages, sections) and only keep the essential in CONTRIBUTING.md with links to the documentation when it's relevant.

**Comment by miri64** [comment]
> This requires that people actually the contributing guidelines before (wanting to) submitting a bug report or feature request. I don't think that's the case.

There is a generated link whenever you open a PR or issue.

image

It used to be much more prominent, but it is their. Also, whenever you type a new comment, there is a link:

image

Do you think people ignore those? If yes? Why should we have a CONTRIBUTING.md at all? 

> Note that I initially rewrote this document in #10029 by taking inspiration from the [rust contributing document](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md). In the meantime, Rust has completely refactored it: their document now points to their related documentation.
> 
> I think we should do the same for RIOT: move details to doc.riot-os.org (in different pages, sections) and only keep the essential in CONTRIBUTING.md with links to the documentation when it's relevant.

If this increases visibility :+1:

**Comment by silkeh** [comment]
It seems there's some consensus to bringing back an issue section. I will restore a shortened version tomorrow which incorporates the feedback here.

> Do you think people ignore those? If yes? Why should we have a CONTRIBUTING.md at all?

Yes. Honestly, I have never even *noticed* the helpful resources in the sidebar of the issue creation screen (and I'm definitely not the only one). But at the point I can see that set of links, I have already made my choice for *bug report*, *feature request* or *security bug report*. At that point there is an (extensive) template, so I wouldn't click on those links unless I need more help.

For PRs this is different: I usually skim the CONTRIBUTING documentation before I start writing code (and a casual glance before submitting the PR). That might be because I'm a drive-by contributor that's used to reading those beforehand, I have no idea how other people approach this. I hope most people don't read the guidelines when their PR is mostly done.

As for the link below the 'Comment' button: I have no idea what purpose that serves.

**Comment by miri64** [comment]
> As for the link below the 'Comment' button: I have no idea what purpose that serves.

We don't have that, but it is there in case there guidelines on commenting in the contributing guidelines. Contributing to a discussion in a comment is also contributing ;-).

**Comment by silkeh** [comment]
I have restored a shorter (but in content mostly the same) version of the bug reports/feature requests sections as a single section. I *think* this covers everything that was mentioned previously.

Additionally, I noticed all the general tips are about contributing code. So I've added another commit moving those to the right section.

File: CONTRIBUTING.md (hunk 1 above)
Comment (silkeh) [reply CONTRIBUTING.md]: Ack. Fixed!

**Comment by silkeh** [comment]
@waehlisch, @aabadie, @miri64: are the changes I made sufficient?

**Review by waehlisch — APPROVED** [review]
looks good to me now.

**Comment by bergzand** [comment]
Please squash!

**Review by miri64 — APPROVED** [review]
ACK from my side as well.

**Comment by silkeh** [comment]
Squashed!

**Pull request merged**